x/dex/keeper/grpc_query_estimate_multi_hop_swap.go

```diff
@@ -28,23 +28,23 @@ func (k Keeper) EstimateMultiHopSwap(
 	ctx := sdk.UnwrapSDKContext(goCtx)
 	cacheCtx, _ := ctx.CacheContext()
 
-	callerAddr := sdk.MustAccAddressFromBech32(req.Creator)
-	receiverAddr := sdk.MustAccAddressFromBech32(req.Receiver)
-
+	oldBk := k.bankKeeper
+	k.bankKeeper = NewSimulationBankKeeper(k.bankKeeper)
```

Comment:
do we do that bank keeper substitution only because of [these module<>bankKeeper interactions](https://github.com/neutron-org/neutron/blob/e57b3d56eb86faacb86f687da57a1fa5d7c7e6e9/x/dex/keeper/core.go#L292-L312)? also I can see that we use a cached context to not update the dex keeper's state. if yes, I think we should rethink the way we implemented the code.

for example, speaking of the [MultiHopSwapCore](https://github.com/neutron-org/neutron/blob/e57b3d56eb86faacb86f687da57a1fa5d7c7e6e9/x/dex/keeper/core.go#L238) method: the logic for a simulation and for a real swap is mostly the same:
**1.** calc the best route
**2.** write route context (mutate dex storage as far as I understand)
**3.** send coins between receiver, caller and module account
**4.** emit events

why don't we create a separate method that does only the **1.** step and is used in both simulation and real swap? in simulation we only call this method and return the result, but in the real swap we also do [the remaining necessary steps](https://github.com/neutron-org/neutron/blob/e57b3d56eb86faacb86f687da57a1fa5d7c7e6e9/x/dex/keeper/core.go#L291-L323)?

Reply:
Agree, i prefer a pure function that does estimation.